packages/react-compose/src/next/getSlots.tsx

```diff
@@ -0,0 +1,52 @@
+import * as React from 'react';
+import { getNativeElementProps, omit } from '@uifabric/utilities';
+import { GenericDictionary } from './types';
+import { nullRender } from './nullRender';
+
+/**
+ * Given the state and an array of slot names, will break out `slots` and `slotProps`
+ * collections.
+ *
+ * The root is always derived from the `as` prop.
+ *
+ * Slots will render as null if they are rendered as primitives with undefined children.
+ *
+ * The slotProps will always omit the `as` prop within them, and for slots that are string
+ * primitives, the props will be filtered according the the slot type. For example, if the
+ * slot is rendered `as: 'a'`, the props will be filtered for acceptable anchor props.
+ *
+ * @param state - State including slot definitions
+ * @param slotNames - Name of which props are slots
+ * @returns An object containing the `slots` map and `slotProps` map.
+ */
+export const getSlots = (state: GenericDictionary, slotNames?: string[] | undefined) => {
+  const slots: GenericDictionary = {
+    root: state.as || nullRender,
+  };
+  const slotProps: GenericDictionary = {
+    root: typeof state.as === 'string' ? getNativeElementProps(state.as, state) : omit(state, { as: 1 }),
+  };
+
+  if (slotNames) {
+    for (const name of slotNames) {
+      const slotDefinition = state[name];
+      const { as: slotAs, children } = slotDefinition;
```

Comment:
This thing implements different behavior than current shorthand API, I mentioned this multiple times. It relies only on `as` prop value which may lead to broken scenarios.

```jsx
  mergeProps(
    {
      image: { as: Image, src: "foo.png", fit: 'center' } // Image supports `fit` prop
    },
    {
      image: { as: 'img' } // Image does not support `fit` prop
    }
  )
```

`mergeProps()` & `getSlots()` will produce `<img src="foo.png" fit="center" />`
`createShorthand()` will produce `<Image as='img' src="foo.png" fit="center" />`

Reply:
Effectively, don't use `as` for the default slot container; rather separate it out into something which the user doesn't stomp on (which is what `createShorthand` is doing.) That way you can use slot containers other than primitives.

I think the analogy here would be to retain a defaultSlots collection which does not get merged. Later when we call `getSlots`, pass that along:

```tsx
const defaultSlots = { image: Image };
const { slots, slotProps } = getSlots(state, defaultSlots);
```

Reply:
@khmakoto @xugao

Reply:
I've opened an RFC to track this discussion here: https://github.com/microsoft/fluentui/issues/14373

We could certainly tweak things to address this case, but I think the current solution is most efficient. We're avoiding hard edges to container components, we don't have both "slots" and "as" to replace the container, and if you really need to render an `Image` with an extra `as` prop, you can do that through the children function.

In your particular example, you started with a default of using the `Image` component. Then the customer or parent component overrode that by using an `img` primitive. Because it's a primitive, getSlots would have slimmed down the props using `getNativeElementProps` and `fit` would have been removed. This is precisely the expected behavior.